core/src/test/java/org/elasticsearch/gateway/GatewayIndexStateIT.java

```diff
@@ -323,6 +323,39 @@ public Settings onNodeStopped(String nodeName) throws Exception {
     }
 
     /**
+     * This test ensures that when an index deletion takes place while a node is offline, when that
+     * node rejoins the cluster, it deletes the index locally instead of importing it as a dangling index.
+     */
+    public void testIndexDeletionWhenNodeRejoins() throws Exception {
+        final String indexName = "test-index-del-on-node-rejoin-idx";
+        logger.info("--> starting two nodes");
+        internalCluster().startNodesAsync(2).get();
+
+        logger.info("--> create an index");
+        createIndex(indexName);
+
+        logger.info("--> waiting for green status");
+        ensureGreen();
+
+        logger.info("--> stop the non-master node");
+        final String stoppedNode = internalCluster().removeRandomNonMasterFromCluster();
+        assertNotNull("should have stopped a non-master node", stoppedNode);
+
+        logger.info("--> delete index and verify it is deleted");
+        client().admin().indices().prepareDelete(indexName).execute().actionGet();
+        assertFalse(client().admin().indices().prepareExists(indexName).execute().actionGet().isExists());
+
+        logger.info("--> restart the stopped node");
+        internalCluster().rejoinNodeToCluster(stoppedNode);
+
+        logger.info("--> waiting for green status");
+        ensureGreen();
+
+        logger.info("--> verify that the deleted index is removed from the restarted node");
+        assertFalse(client().admin().indices().prepareExists(indexName).execute().actionGet().isExists());
```

Comment:
can we also check that the index folder on disk is cleaned, via org.elasticsearch.env.NodeEnvironment#availableIndexFolders?